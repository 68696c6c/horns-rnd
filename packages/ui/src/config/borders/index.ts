import { SizeOption } from '../sizes'

export enum BorderStyle {
  None = 'none',
  Hidden = 'hidden',
  Dotted = 'dotted',
  Dashed = 'dashed',
  Solid = 'solid',
  Double = 'double',
  Groove = 'groove',
  Ridge = 'ridge',
  Inset = 'inset',
  Outset = 'outset',
  Initial = 'initial',
  Inherit = 'inherit',
}

export type BorderStyleOption = `${BorderStyle}`

export type BorderProperties = {
  width?: SizeOption
  style?: BorderStyleOption
}

export type BordersConfig = {
  all?: BorderProperties
  x?: BorderProperties
  y?: BorderProperties
  top?: BorderProperties
  bottom?: BorderProperties
  left?: BorderProperties
  right?: BorderProperties
}
