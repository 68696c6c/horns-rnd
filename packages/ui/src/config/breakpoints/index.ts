import _merge from 'lodash.merge'

export enum Breakpoint {
  Mobile = 'mobile',
  Min = 'min',
  Small = 'small',
  Medium = 'medium',
  Large = 'large',
  Max = 'max',
}

export type BreakpointsConfig = {
  [key in Breakpoint]: string
}

export const defaultConfig: BreakpointsConfig = {
  mobile: 'small',
  min: '320px',
  small: '480px',
  medium: '768px',
  large: '992px',
  max: '1200px',
}

export class Breakpoints {
  mobile: string

  min: string

  small: string

  medium: string

  large: string

  max: string

  constructor(config?: Partial<BreakpointsConfig>) {
    const mergedConfig = _merge(defaultConfig, config)

    this.mobile = mergedConfig[mergedConfig.mobile as Breakpoint]
    this.min = mergedConfig.min
    this.small = mergedConfig.small
    this.medium = mergedConfig.medium
    this.large = mergedConfig.large
    this.max = mergedConfig.max
  }
}
