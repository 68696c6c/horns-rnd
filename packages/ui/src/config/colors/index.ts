import _merge from 'lodash.merge'

import { DeepPartial } from '../utils'
import {
  Mode,
  Config,
  defaultConfig,
  makePallet,
  Colorways,
  makeColorways,
  Backgrounds,
  makeBackgrounds,
} from './types'

export { Colorway, Background } from './types'

export type ColorsConfig = DeepPartial<Config>

export interface Colors {
  readonly isDarkMode: boolean
  readonly colorways: Colorways
  readonly backgrounds: Backgrounds
}

export const makeColors = (config?: ColorsConfig) => {
  const mergedConfig = _merge(defaultConfig, config)
  const pallet = makePallet(mergedConfig)
  return {
    isDarkMode: mergedConfig.mode === Mode.Dark,
    colorways: makeColorways(pallet, mergedConfig),
    backgrounds: makeBackgrounds(pallet, mergedConfig),
  }
}
