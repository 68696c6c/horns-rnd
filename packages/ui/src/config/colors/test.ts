import Color from 'color'

import { makeColors } from '.'

describe('Colors', () => {
  it('should use dark background colors if mode is set to "dark"', () => {
    const config = makeColors({ mode: 'dark' })
    const result = config.backgrounds.primary.base.base
    expect(Color(result).isDark()).toBe(true)
  })

  it('should use light background colors if mode is set to "light"', () => {
    const config = makeColors({ mode: 'light' })
    const result = config.backgrounds.primary.base.base
    expect(Color(result).isDark()).toBe(false)
  })
})
