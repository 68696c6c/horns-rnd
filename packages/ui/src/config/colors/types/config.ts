import Color from 'color'

import { Colorway } from './types'

export enum Mode {
  Light = 'light',
  Dark = 'dark',
}

export type PalletConfig = {
  [key in Colorway]: string
}

// These template literal types exist to allow users to provide config values as strings instead of
// enums since they won't have access to our enums.
type ModeOption = `${Mode}`
type ProminentOption = `${Colorway}`

export interface Config {
  mode: ModeOption
  prominent: ProminentOption
  pallet: Omit<PalletConfig, 'prominent' | 'background' | 'backgroundAlt'>
  shaders: Shaders
}

export interface Shaders {
  alpha: number
  darkest: number
  lightest: number
  darker: MinMax
  dark: MinMax
  light: MinMax
  lighter: MinMax
}

export interface MinMax {
  min: number
  max: number
}

export const hexPallet = {
  violet: '#7f00ff',
  indigo: '#3f00ff',
  blue: '#1a99ff',
  cyan: '#17a2b8',
  teal: '#009999',
  green: '#28a745',
  yellow: '#ffc107',
  tangerine: '#ffaa00',
  mustard: '#aa7700',
  orange: '#fd7e14',
  red: '#dc3545',
  white: '#ffffff',
  gray: '#7f7f7f',
  black: '#010101',
}

export const colorPallet = {
  // violet: Color(hexPallet.violet),
  // indigo: Color(hexPallet.indigo),
  // blue: Color(hexPallet.blue),
  // cyan: Color(hexPallet.cyan),
  // teal: Color(hexPallet.teal),
  // green: Color(hexPallet.green),
  // yellow: Color(hexPallet.yellow),
  // tangerine: Color(hexPallet.tangerine),
  // mustard: Color(hexPallet.mustard),
  // orange: Color(hexPallet.orange),
  // red: Color(hexPallet.red),
  white: Color(hexPallet.white),
  gray: Color(hexPallet.gray),
  black: Color(hexPallet.black),
}

export const defaultConfig: Config = {
  mode: 'light',
  prominent: 'primary',
  pallet: {
    primary: hexPallet.tangerine,
    secondary: hexPallet.blue,
    tertiary: hexPallet.teal,
    dark: hexPallet.black,
    neutral: hexPallet.gray,
    light: hexPallet.white,
    // dark: '#070706',
    // neutral: '#887E6A',
    // light: '#F9F9F8',
    success: hexPallet.green,
    info: hexPallet.cyan,
    warning: hexPallet.yellow,
    danger: hexPallet.red,
  },
  shaders: {
    alpha: 0.3,
    darkest: 88,
    lightest: 20,
    darker: {
      min: 0.08,
      max: 0.155,
    },
    dark: {
      min: 0.2,
      max: 0.47,
    },
    light: {
      min: 0.22,
      max: 0.55,
    },
    lighter: {
      min: 25,
      max: 40,
    },
  },
}
