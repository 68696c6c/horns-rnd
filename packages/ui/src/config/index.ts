export { ButtonsConfig, defaultButtons } from './buttons'
export { BorderStyle, BordersConfig, BorderProperties } from './borders'
export { Breakpoint } from './breakpoints'
export { Colorway, Background } from './colors'
export { Cursor } from './cursors'
export { Theme, Config } from './theme'
export { Size, SizeOption, SideSizeOptions } from './sizes'
export { Font } from './typography'
export { SideOption, Side, UiState, HoverState, StatusState } from './utils'
