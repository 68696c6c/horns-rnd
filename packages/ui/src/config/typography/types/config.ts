import { UiState } from '../../utils'

import { defaultDecorations, Decoration, Decorations } from './decorations'
import { defaultFontFamilies, FontFamilies, FontFamily } from './families'
import { Font, FontConfig } from './fonts'
import { defaultLettings, Letting, Lettings } from './lettings'
import { defaultFontSizes, BaseFontSize, FontSizes } from './sizes'
import { defaultTextSpacings, TextSpacing, TextSpacings } from './spacings'
import { defaultFontWeights, FontWeight, FontWeights } from './weights'
import {
  Direction,
  FontKerning,
  FontStyle as FontStyleCss,
  FontStyle,
  TextAlign,
  TextTransform,
} from './types'

export type ConfigFontStyles = {
  [key in Font]: ConfigFontStates
}

export type ConfigFontStates = {
  [key in UiState]?: Partial<FontConfig>
}

export interface Config {
  direction: Direction.LTR
  families: FontFamilies
  decorations: Decorations
  weights: FontWeights
  sizes: FontSizes
  spacing: TextSpacings
  letting: Lettings
  styles: ConfigFontStyles
}

export const makeDefaultFontConfig = (config: Config): FontConfig => ({
  family: FontFamily.Primary,
  style: FontStyleCss.Normal,
  weight: FontWeight.Base,
  size: BaseFontSize.Base,
  align: config.direction === 'ltr' ? TextAlign.Left : TextAlign.Right,
  transform: TextTransform.None,
  decoration: Decoration.Base,
  kerning: FontKerning.Normal,
  spacing: TextSpacing.Base,
  indent: '',
  letting: Letting.Base,
  tracking: '',
})

export const defaultConfig: Config = {
  direction: Direction.LTR,
  families: defaultFontFamilies,
  decorations: defaultDecorations,
  weights: defaultFontWeights,
  sizes: defaultFontSizes,
  spacing: defaultTextSpacings,
  letting: defaultLettings,
  styles: {
    heading: {
      base: {
        weight: FontWeight.Bold,
        spacing: TextSpacing.Max,
      },
    },
    paragraph: {
      base: {
        align: TextAlign.Justify,
        spacing: TextSpacing.Max,
      },
    },
    quote: {
      base: {
        style: FontStyle.Italic,
      },
    },
    text: {},
    table: {
      base: {
        letting: Letting.Min,
      },
    },
    nav: {
      base: {
        letting: Letting.Min,
      },
    },
    control: {
      base: {
        letting: Letting.Min,
      },
    },
    label: {
      base: {
        weight: FontWeight.Bold,
      },
    },
    message: {
      base: {
        letting: Letting.Min,
      },
    },
    button: {
      base: {
        weight: FontWeight.Bold,
      },
    },
    link: {
      base: {
        decoration: Decoration.Underline,
      },
      hover: {
        style: FontStyle.Italic,
        decoration: Decoration.Underline,
      },
      active: {
        style: FontStyle.Italic,
        decoration: Decoration.UnderlineDouble,
      },
      inactive: {
        decoration: Decoration.Underline,
      },
    },
    caption: {
      base: {
        weight: FontWeight.Light,
      },
    },
    legal: {
      base: {
        family: FontFamily.Secondary,
        style: FontStyle.Italic,
        size: BaseFontSize.Small,
      },
    },
    code: {
      base: {
        family: FontFamily.Tertiary,
      },
    },
    emphasized: {
      base: {
        style: FontStyle.Italic,
      },
    },
    strong: {
      base: {
        weight: FontWeight.Light,
      },
    },
    variable: {
      base: {
        family: FontFamily.Tertiary,
        style: FontStyle.Italic,
      },
    },
    mistake: {
      base: {
        decoration: Decoration.Mistake,
      },
    },
  },
}
