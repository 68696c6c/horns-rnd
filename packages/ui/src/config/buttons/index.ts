import { BordersConfig } from '../borders'
import { Cursor } from '../cursors'
import { Font } from '../typography'
import { SideSizeOptions } from '../sizes'

export type ButtonsConfig = {
  border: BordersConfig
  cursor: Cursor
  font: Font
  padding: SideSizeOptions
}

export const defaultButtons = {
  border: {
    all: {
      width: 'tiny',
      style: 'solid',
    },
  },
  cursor: 'pointer',
  font: 'button',
  padding: {
    x: 'medium',
    y: 'small',
  },
}
