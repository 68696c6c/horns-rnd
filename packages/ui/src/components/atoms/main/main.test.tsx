/* eslint-disable react/jsx-one-expression-per-line */
import React from 'react'

import { render } from '../../test'
import { Colorway, Size } from '../../../config'

import { Main } from '.'

describe('Main', () => {
  it('should render as default', () => {
    const { container } = render(
      <Main>
        <h1>section</h1>
        <p>content can go anywhere</p>
        <div>a div</div>
        <div>another div</div>
      </Main>,
    )
    expect(container).toMatchSnapshot()
  })
  it.each(Object.values(Colorway))('should render colorway %s', (color) => {
    const { container } = render(
      <Main color={color}>
        <h1>{color} section</h1>
        <p>{color} content</p>
      </Main>,
    )
    expect(container).toMatchSnapshot()
  })
  it.each(Object.values(Size))('should render padding %s', (size) => {
    const { container } = render(<Main padding={size}>{size}</Main>)
    expect(container).toMatchSnapshot()
  })
})
