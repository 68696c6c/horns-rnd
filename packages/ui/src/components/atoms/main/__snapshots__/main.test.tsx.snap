// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main should render as default 1`] = `
.emotion-0 {
  border-color: rgb(224, 224, 224);
  background-color: rgb(255, 255, 255);
  color: rgb(1, 1, 1);
}

@media (min-width: 1200px) {
  .emotion-0 {
    padding-left: calc(((100vw - 1200px) / 2));
    padding-right: calc(((100vw - 1200px) / 2));
  }
}

<div>
  <main
    class="emotion-0"
  >
    example
  </main>
</div>
`;

exports[`Main should support trait props 1`] = `
.emotion-0 {
  border-color: rgb(204, 136, 0);
  background-color: rgb(255, 170, 0);
  color: rgb(1, 1, 1);
  padding-top: 12px;
  padding-right: 12px;
  padding-bottom: 12px;
  padding-left: 12px;
}

@media (min-width: 1200px) {
  .emotion-0 {
    padding-left: calc(((100vw - 1200px) / 2));
    padding-right: calc(((100vw - 1200px) / 2));
  }
}

<div>
  <main
    class="emotion-0"
    color="primary"
  >
    example
  </main>
</div>
`;
